core/templates/dev/head/pages/exploration_editor/ExplorationEditor.js

```diff
@@ -554,443 +539,47 @@ oppia.controller('ExplorationSaveAndPublishButtons', [
-            directionForwards: true
-          }]);
-        $scope.diffData = {
-          nodes: diffGraphData.nodes,
-          links: diffGraphData.links,
-          finalStateIds: diffGraphData.finalStateIds,
-          v1InitStateId: diffGraphData.originalStateIds[data.init_state_name],
-          v2InitStateId: diffGraphData.stateIds[
-            explorationInitStateNameService.displayed],
-          v1States: oldStates,
-          v2States: newStates
-        };
-
-        // TODO(wxy): after diff supports exploration metadata, add a check to
-        // exit if changes cancel each other out.
-
-        alertsService.clearWarnings();
-
-        // If the modal is open, do not open another one.
-        if (_modalIsOpen) {
-          return;
-        }
-
-        var modalInstance = $modal.open({
-          templateUrl: 'modals/saveExploration',
-          backdrop: true,
-          resolve: {
-            isExplorationPrivate: function() {
-              return explorationRightsService.isPrivate();
-            },
-            diffData: function() {
-              return $scope.diffData;
-            }
-          },
-          windowClass: 'oppia-save-exploration-modal',
-          controller: [
-            '$scope', '$modalInstance', 'isExplorationPrivate', 'diffData',
-            function($scope, $modalInstance, isExplorationPrivate, diffData) {
-              $scope.showDiff = false;
-              $scope.onClickToggleDiffButton = function() {
-                $scope.showDiff = !$scope.showDiff;
-                if ($scope.showDiff) {
-                  $('.oppia-save-exploration-modal').addClass(
-                    'oppia-save-exploration-wide-modal');
-                } else {
-                  $('.oppia-save-exploration-modal').removeClass(
-                    'oppia-save-exploration-wide-modal');
-                }
-              };
-
-              $scope.diffData = diffData;
-              $scope.isExplorationPrivate = isExplorationPrivate;
-
-              $scope.earlierVersionHeader = 'Last saved';
-              $scope.laterVersionHeader = 'New changes';
-
-              $scope.save = function(commitMessage) {
-                $modalInstance.close(commitMessage);
-              };
-              $scope.cancel = function() {
-                $modalInstance.dismiss('cancel');
-                alertsService.clearWarnings();
-              };
-            }
-          ]
-        });
-
-        // Modal is Opened
-        _modalIsOpen = true;
-        $scope.saveModalIsOpening = false;
+    $scope.isSaveModalOpening = function() {
+      return explorationSaveService.isSaveModalOpening();
+    };
 
-        modalInstance.opened.then(function() {
-          // The $timeout seems to be needed in order to give the modal time to
-          // render.
-          $timeout(function() {
-            focusService.setFocus('saveChangesModalOpened');
-          });
-        });
+    $scope.diffData = function() {
```

Comment:
Why do you need to expose this? It's only shown in the modal, AFAICT.

In fact, anything specific to the modal should not be exposed in this controller. So, in this service, I think you want:
- isPrivate()
- isExplorationPublishable()
- isExplorationSaveable()
- getPublishExplorationButtonTooltip()
- getSaveButtonTooltip()
- getChangeListLength()
- isEditableOutsideTutorialMode()

and local $scope-bound vars for publishModalIsOpening, saveIsInProgress, saveModalIsOpening. (I would suggest using vars if possible to keep track of state.)

This gives you a nice separation between stuff you need to display the save and publish buttons themselves (which is the point of this controller), and all the other core logic that can be reused elsewhere.